### ISSUE TITLE
Support for appdynamics

### ISSUE THREAD

**Issue by bjartek** [micrometer-metrics/micrometer#575]
It is possible to hook micrometer into https://www.appdynamics.com/ in some way?

Relevant links from slack 
https://docs.appdynamics.com/display/PRO41/Build+a+Monitoring+Extension+Using+Java

https://docs.appdynamics.com/display/PRO42/Analytics+Events+API#AnalyticsEventsAPI-AbouttheAnalyticsEventsAPI

**Comment by checketts** [comment]
I'll be working with a company that uses AppDynamics shortly and should be able to dig into this.

**Comment by bjartek** [comment]
Has there been any progress on this issue?

**Comment by jkschneider** [comment]
@checketts Is there anything we can do from a partner perspective to reach out and ask AppDynamics to publish a jar to maven central?

**Comment by checketts** [comment]
I'm pretty disappointed with AppDynamics. It has a very hierarchical metrics concept, so I've paused on it and instead have been working toward Splunk metrics support via HEC (http event collector).

Progress has been slow on that one too, but I am making progress when I have time to work on it.

**Comment by jkschneider** [comment]
I was looking at AQDL from the outside in today and it looks like you might be able to select dimensions using WHERE clauses. Combining this with [analytics functions](https://docs.appdynamics.com/display/PRO44/Analytics+Functions) seems like it _could_ work. Did you go down the same path?

**Comment by checketts** [comment]
No, I didn't look into that part. I didn't know it had a query language.

I was just ramping up on AppD and trying to get the metrics in when I decided to switch focus.

**Comment by bjartek** [comment]
Any progress on this?

**Comment by checketts** [comment]
No I've tabled my work on this for now.

**Comment by jkschneider** [comment]
@bjartek I did have a chat with an AppD engineer at Oracle CodeOne recently. Looks like we'll essentially have two implementations: a hierarchical one passing through the Java agent and a dimensional one going to the Analytics Event API. Evidently, licenses for the agent-based approach are separate from licenses for the Analytics Event API.

**Comment by febinrejoe** [comment]
Any update on this issue?

**Comment by shakuzen** [comment]
> Any update on this issue?

Only the information available as comments on this issue. Which license do users requesting this have so we can know what solution would work for them? Another solution would be for AppDynamics customers to request AppDynamics to support collecting Micrometer metrics from the AppDynamics Java agent.
Do users want Micrometer metrics in AppDynamics for the purpose of custom metrics in their application or because of some metrics Micrometer provides but AppDynamics does not?

**Comment by dermoritz** [comment]
I just stumbled across this. For me the main use case is, that you can develop your metrics or insights independently of an metric application like AppDynamics.
It is like logging - nowadays no one uses a specific log-Implementation but Slf4J.

**Comment by checketts** [comment]
@dermoritz Are you saying you are interested in contributing AppDynamics support? I wouldbe happy to mentor the effort.

**Comment by rjmveloso** [comment]
I came across this issue when searching for a micrometer exporter to AppDynamics.
Despite all the magic that AppDynamics provides out-of-the-box this integration could be handy for custom metrics.
I've been working in an monitor extension to retrieve statistics from the actuator endpoint: https://github.com/rjmveloso/actuator-monitoring-extension
However I think a better solution could be a micrometer exporter that makes use of the machine agent rest API (if enabled).

**Comment by ameyg7** [comment]
Team,
What is the update on this?
Do we have full support for AppDynamics integration?
Which all features of AppD can be integrated with Micrometer?

**Comment by checketts** [comment]
No changes that I am aware of. There is no AppD exporter yet.

**Comment by jonatan-ivanov** [comment]
It seems though that AppD has support for Prometheus, StatsD, and Elasticsearch but for these you will need extra infrastructure.

**Comment by rjmveloso** [comment]
If possible to use a monitoring extension, I believe my implementation is a good starting point.
I will provide shortly a draft for an AppDynamicsMeterRegistry. I will be glad if anyone could help on contribution.
By now, one of my main concerns is providing the min value required by AppD metrics. I’m following the same approach used by the implementation of Dynatrace, with a DistributionStatisticConfig with the percentile 0 for that purpose.
Need some time to get it working and tested.

**Comment by shakuzen** [comment]
We appreciate the offer for a contribution @rjmveloso. Let us know if you have something for us to take a look at. I'm not sure if we have access to AppDynamics for the Micrometer team to test. We can check into that.
Another option that may be available at some point is that I see OpenTelemetry Protocol support in AppDynamics for traces now. If they add support for metrics, you could try our micrometer-registry-otlp module for sending metrics to AppDynamics via OTLP format.

**Comment by rjmveloso** [comment]
Hello all, a draft approach can be found at https://github.com/rjmveloso/micrometer-registry-appdynamics
Please feel free to comment, contribute, fork, as you pleased.

AppDynamics Agent API documented that already aggregated metrics may be reported (which is the case with Micrometer) but requires a min value. As stated, this point has to be implemented since Micrometer does not keep min values.
A recent investigation on Dynatrace implementation, they have replaced the "artifical" zero percentile with a custom DistributionSummary that keeps track of the min value. I will try to implement the same principle in the following days. Currently min value is being reported as "0".

**Comment by rjmveloso** [comment]
Hello everyone,
Implementation is done and it was tested with our own proper cases.
Anyone who was willing for this would please have a try? @ameyg7 , @bjartek , @dermoritz maybe.
@shakuzen before propose the implementation to Micrometer is it possible for you to have a look in my repo?
